File: done/Q1260.py
#dfs = stack
#bfs = queue
#재귀보다는 이렇게 구현하자

n, m, v = map(int, input().split())
#map은 리스트의 요소를 지정된 함수로 처리해줌/새 리스트를 만듬?
#n, m, v에 각각 처음 input을 split한 값을 순서대로 int 형태로 넣어줌
#map을 사용하면 한번에 형지정, 변수 할당이 가능함

#정석의 그래프 만드는 것은 아래와 같다(연결 리스트 만들기)

'''
#만약 n개만큼의 정점이 존재하고, m개만큼의 입력을 받는다면 
graph = [[] for _ in range(n+1)] # n+1개만큼의 공간을 만들어서 graph[n]이 n번 정점을 나타내도록 해 준다 
for _ in range(m):
	x,y = map(int,input().split()) #만약 1 2를 입력받으면 
    graph[x].append(y) # 1번 정점에 2를 넣어주고 -> graph[1] = [2] 
    graph[y].append(x) # 2번 정점에 1을 넣어준다 -> graph[2] = [1]
'''

#Graph set 중복 거르고 교집합 구하기 쉽게 set 사용(위 기본식에서 약간 변형)
graph_list = [set([]) for _ in range(n+1)]

for _ in range(m):
    i, j = map(int, input().split())
    graph_list[i].add(j)
    graph_list[j].add(i)

#DFS
def dfs(graph_list, start):
    visited = []
    stack = [start]

    while stack:
        node = stack.pop()
        if node not in visited:
            visited.append(node)
            stack += sorted(list(graph_list[node] - set(visited)), reverse = True)
            #스택은 뒤에서 나가니 역순 정렬
    return visited

#BFS
def bfs(graph_list, start):
    visited = []
    queue = [start]
    while queue:
        node = queue.pop(0)
        if node not in visited:
            visited.append(node)
            queue += sorted(list(graph_list[node] - set(visited)))
            #큐는 앞부터 나가니 순정렬
    return visited

for i in list(dfs(graph_list, v)):
    print(i, end=" ")
print()
for j in list(bfs(graph_list, v)):
    print(j, end=" ")

File: done/Q1260.java
package done;

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.*;

public class Q1260 {

	static int N, M;
    static ArrayList<Integer>[] nodeList;
    static boolean[] isVisted;
    
    static StringBuilder sb = new StringBuilder();

    public static void main(String[] args) throws Exception {
    	BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    	String inputStr = br.readLine();
    	String array[] = inputStr.split(" ");
    	
        N = Integer.parseInt(array[0]);    // 정점 수
        M = Integer.parseInt(array[1]);    // 간선 수
        int V = Integer.parseInt(array[2]);    // 시작 정점

        nodeList = new ArrayList[N+1];    // 정점수대로 배열 생성
        isVisted = new boolean[N+1];         // 방문 여부를 판단

        // 간선 리스트 초기화
        // 1번 정점부터 리스트를 만들어서 배열에 넣기
        for (int i = 1; i <= N; i++) {
            nodeList[i] = new ArrayList<>();
        }

        // 간선수대로 반복하며 연결 리스트 생성
        for (int i = 0; i < M; i++) {
        	String edgeStr = br.readLine();
        	String edgeArray[] = edgeStr.split(" ");
            int from = Integer.parseInt(edgeArray[0]);
            int to = Integer.parseInt(edgeArray[1]);

            // 양방향
            nodeList[from].add(to);
            nodeList[to].add(from);
        }

        DFS(V);

        Arrays.fill(isVisted, false);
        sb.append("\n");

        BFS(V);

        System.out.print(sb);
    }

    private static void DFS(int V) {
        if(isVisted[V]) return;

        sb.append(V).append(" ");
        isVisted[V] = true;

        // 해당 노드에 존재하는 노드 수만큼 반복
        for (int i = 0; i < nodeList[V].size(); i++) {
            int MIN = 1001;

            // 해당 노드가 가장 낮은 수인지 모두 방문하여 검증
            for (int j = 0; j < nodeList[V].size(); j++) {
                int cur = nodeList[V].get(j);

                if(!isVisted[cur]) {
                    MIN = Math.min(MIN, cur);
                }
            }

            if(MIN == 1001) break;

            DFS(MIN);
        }
    }

    private static void BFS(int V) {
        Queue<Integer> q = new LinkedList<>();

        q.add(V);
        isVisted[V] = true;

        while(!q.isEmpty()) {
            int node = q.poll();
            sb.append(node).append(" ");

            // 해당 노드에 존재하는 노드 수만큼 반복
            for (int i = 0; i < nodeList[node].size(); i++) {
                int MIN = 1001;

                // 해당 노드가 가장 낮은 수인지 모두 방문하여 검증
                for (int j = 0; j < nodeList[node].size(); j++) {
                    int cur = nodeList[node].get(j);

                    if(!isVisted[cur]) {
                        MIN = Math.min(MIN, cur);
                    }

                }

                if(MIN == 1001) break;

                q.add(MIN);
                isVisted[MIN] = true;
            }

        }

    }

}
